### PR TITLE
Decode protected emails

### LINKED ISSUE #41: Email protection is not parsed by cloudscraper
On Crunchyroll, they have activated the "email protection" from cloudflare and when I try to scrape some of their webpages I get "[email protected]" instead of the expected text.

Example: http://www.crunchyroll.com/the-idolmster-cinderella-girls-theater#

On that webpage the "DOLM@STER" is seen as an email by cloudflare (that mean their regexp is fabulously wrong) and it screw my scraping because I'm not expecting ton of the javascript they insert for that "protection". 

It would be nice if cloudscraper could detect these insertion and parse them the same as it does for the browser detection.

### PR DESCRIPTION
**Pull request by pro-src**
Closes #41 

TODO:
- [x] Add tests

### CODE REVIEW

**Comment by pro-src** [comment]
@codemanki I didn't go for a flag. It seemed to over complicate things. Instead, I added a decode method. If you think a flag should be added to this PR, just lemme know.

```js
cloudcraper.get('http://example-site.dev', (error, response, body) =>  {
  if (!error) {
    body = cloudscraper.decode(body);
  }
});
```

**Comment by codemanki** [comment]
@pro-src I'm not sure if i actually want to have this functionality in here. This now looks a bit fishy and we give spammers more opportunities to annoy users who try to avoid that. What is your opinion on this?

**Comment by pro-src** [comment]
@codemanki I don't think the majority of spammers are hindered by Cloudflare's email protection. I found that many developers have published code to bypass this in many languages. Cloudflare isn't concerned so why should we be? If cloudscraper gets enough attention for this feature, Cloudflare might improve their protection which is not good for spammers. And lastly, there are legitimate use cases.

**Comment by pro-src** [comment]
I'm quoting from http://cryptographybuzz.com/cloudflares-email-address-obfuscation/

You might have noticed that this encryption strategy is super weak. Storing
the key right next to the ciphertext is barely better than just sending the
email address in plaintext, and a single byte XOR is trivial to detect and brute
force—in fact, it’s the
third exercise of the excellent
Cryptopals challenge.

Indeed, the encoding method isn’t designed to securely encrypt email
addresses: while cryptographically weak, it’s enough to throw off the basic
scripts that hunt for `mailto:` links. *One Cloudflare security engineer wrote:*

> The scrape shield is designed to prevent low-level bots from crawling web pages for contact information. Although it is possible to reveal email addresses due to weak encryption, we do not consider this to be a significant issue. The feature is meant to obfuscate email addresses; not completely enforce their confidentiality. As the alternative would be to not use the scrape shield and display the emails in plaintext, we are of the opinion that this feature does not introduce a vulnerability.

### Prior art
It turns out that many people have done this sort of thing before:

* Raddle user [sudo](https://raddle.me/user/sudo) posted this C++ program
Usama Ejaz wrote deobfuscation routines in [six (!) languages](https://usamaejaz.com/cloudflare-email-decoding/)
SaltyCrane used a [JS interpreter to decode email addresses](https://www.saltycrane.com/blog/2015/07/calling-javascript-python-de-cloudflare-scraped-content/)
/u/ck3k wrote a [routine](https://www.reddit.com/r/netsec/comments/8lduwm/an_analysis_of_cloudflares_email_address/dzfql71/) in Crystal Lang earlier this year for Nettis. Very cool!
/u/nullableVoidPtr did this in a [Python script](https://gist.github.com/nullableVoidPtr/632583d4e34b4b1bf3b92f5b4f5d0c7d)

**Comment by codemanki** [comment]
@pro-src but why would we want to help spammers even if email protection is not perfect? If some wants to scrape emails from webpages they will most likely use it for spamming. I can't find any package doing that, only snippets you mentioned. IMO i'm hesitating if this should be included in cloudscraper

**Comment by pro-src** [comment]
@codemanki We don't want to help them. Cloudflare's security engineers have had plenty of time to address this issue and they're not concerned. You would help the spammers by hiding how easy it is to defeat the email protection. Understand that vulnerabilities should be published. It's the whitehat thing to do and is probably why you don't already see a packages on NPM doing it. Alas, feel free to close this PR and the 2 year old issue if you want. It's your choice but I don't really think it matters that much either way.

**Comment by codemanki** [comment]
@pro-src okay, let's see how it goes :) Hope they won't ask me to take down the package because of this

**Comment by pro-src** [comment]
@codemanki If concerned is raised at Cloudlare, they would know that targeting this package would not solve the problem. They would have to update the email protection which is a good thing. Also, have you ever wondered why the JS challenge is so easy to defeat? Why aren't they using the bot-filter that is blocking us from adding captcha support by default instead? I can only speculate that they're intentionally holding back. If it were an issue, they are more than capable of an upgrade.

File: lib/email-decode.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    codePoint = parseInt(hexStr.substr(i, 2), 16) ^ key;
+    email += String.fromCharCode(codePoint);
+  }
+
+  // noinspection JSDeprecatedSymbols
+  return decodeURIComponent(escape(email));
+}
+
+/*
+function encode(str) {
```
Comment (codemanki) [inline lib/email-decode.js]: Maybe we should remove this code?

**Comment by codemanki** [comment]
```
cloudcraper.get('http://example-site.dev', (error, response, body) =>  {
  if (!error) {
    body = cloudscraper.decode(body);
  }
});
```
Hrm. How do i know that `error` happened because of the protected email? Maybe we should have a separate error for this case?
```
if error typeOf errors.ProtectedEmailError {
    body = cloudscraper.decode(body);
}
```

File: lib/email-decode.js (hunk 1 above)
Comment (pro-src) [reply lib/email-decode.js]: Sure, It's just there for information.

**Comment by pro-src** [comment]
It should never throw an error unless they pass something that cannot be coerced into String. Either the pattern matches or it doesn't. If they need to verify whether or not protection was detected and replaced, they can compare the return value. `if (body === cloudscraper.decode(body)) ...`

**Comment by pro-src** [comment]
Sorry if I misunderstood your question? Are you suggesting that we throw an error if we detect email protection so they know when to decode it? If so,  I think that might be redundant since you'd have to search the document twice. Once to detect and again to remove the protection. I didn't really like the idea of running this for every response that is html because it makes it seems like we're assuming every request is related to Cloudflare.

**Comment by codemanki** [comment]
🤔 i don't get the api then. If some error occurs, user should somehow guess that this is email decode error, run the decode and hope that the result body actually now contains the proper emails? Why not to always decode it regardless of a double scan?

**Comment by pro-src** [comment]
@codemanki compare it to similar APIs and you'll get it. There are entities in the page that should be decoded. If the pattern matches, they're decoded. If it doesn't, no replacements are performed. It's a simple and clean API in that it's not doing anything more than what is necessary (Unix Philosophy). Why would it work any different from https://www.npmjs.com/package/html-entities ? Take a look at that code base https://github.com/mdevils/node-html-entities/blob/master/lib/html5-entities.js

As far always scanning the response if the content-type is text/html, I would prefer to let the user decide if they even need this feature and when it should be used. Should a html-parser force entity decoding on it's users? A method in this case provides the most flexibility.

Concerning errors, there is no way of knowing whether emails should even be in the response so if we wanted to check for errors, we can't. How can you tell the difference between a page that has upgraded email protection and one that doesn't have any emails at all? You can't unless you already know that page contains emails and that's all the more reason for this to be under the user's control.

**Comment by pro-src** [comment]
If you're requesting a flag, I'll gladly add one to this PR :smile:

**Comment by codemanki** [comment]
@pro-src First of all thank you for such a detailed explanation. I totally get your points, but i just realized that i'm probably misunderstanding this email protection feature. Do you have an example of a page with any email protected by CF?

**Comment by codemanki** [comment]
I'm still a bit confused by this functionality. 

We are giving a way for a user to decode any page to extract emails, but no other way of seeing that this protection is present in the page. You are right that if user **knows** that  protected email is present in the page, and they target it, then they will go to the README and see how `decode` method works. But this feels to me like a hidden feature somehow :)

(I don't really want to block this PR, as I have never used CF in this way so i'm not a target audience)

**Comment by pro-src** [comment]
>  Do you have an example of a page with any email protected by CF?

The OP of #41, that this PR closes, has linked a page where this feature is useful.

> I'm still a bit confused by this functionality.

*Only one of the 4 commits in this PR introduces the email decoding feature.*
We can look at this commit specifically. It includes tests which show exactly how it works: https://github.com/codemanki/cloudscraper/pull/146/commits/9361a945391eac31f1426fe075aed1dfe8b39d1a

You can modify the tests to see more clearly what is being done.
```js
  it('should replace anchors that have a data-cfemail attribute', function () {
    var protection = '<a href="/cdn-cgi/l/email-protection" class="__cf_email__" data-cfemail="' +
      HEX_STRING + '">[email&#160;protected]</a>';

    var raw = genHTML('<p> The email is ' + EMAIL + '</p>');
    var enc = genHTML('<p> The email is ' + protection + '</p>');

    expect(decode(enc)).to.equal(raw);
    // Example
    console.log('The page with a protected email:\n ' + enc + '\n\n');
    console.log('The page after decoding:\n' + raw + '\n\n');
  });
```

You can skip/disable a specific test:
```js
it.skip('should not run this test', function(){});
```

Also you can run one specific test instead of all of them:
```js
it.only('should replace anchors that have a data-cfemail attribute', function () {});
```

The other commits in this PR are included to avoid rebasing:
* (PR #144 closed in preference of PR #145 ) https://github.com/codemanki/cloudscraper/pull/146/commits/2319b7845b3f9f2b79d98f521681b912a9c020c2 Replace uri.hostname with uri.host in answer url
* (PR #145 open) https://github.com/codemanki/cloudscraper/pull/146/commits/6c6e96b5fcb74b3267cb6275b214214919b453a2 Superior tests
* (PR #145 open) https://github.com/codemanki/cloudscraper/pull/146/commits/8b7d141de04ac6a61aefe6efd7ca1d1f203712b7 Prevent silent failures

> We are giving a way for a user to decode any page to extract emails, but no other way of seeing that this protection is present in the page.

If they're not aware of the protection in the page then I doubt they need this feature at all.

> But this feels to me like a hidden feature somehow :)

IMHO, if it's in the README then it's not hidden unless the README is just too big or not organized properly. If it does seem hidden in the README then it's a documentation issue.

I can make changes if you request them. e.g. I could add a flag that would allow users to enable this by default.

**Comment by pro-src** [comment]
@codemanki PR #145 includes a bug fix. Can we focus on that and come back to this later?

**Comment by pro-src** [comment]
@codemanki I've opened a new PR #159

**Pull request closed without merging**